stdlib/functools.pyi

```diff
@@ -42,8 +46,17 @@ else:
 WRAPPER_ASSIGNMENTS: Sequence[str]
 WRAPPER_UPDATES: Sequence[str]
 
-def update_wrapper(wrapper: _T, wrapped: _AnyCallable, assigned: Sequence[str] = ..., updated: Sequence[str] = ...) -> _T: ...
-def wraps(wrapped: _AnyCallable, assigned: Sequence[str] = ..., updated: Sequence[str] = ...) -> Callable[[_T], _T]: ...
+class _Wrapped(Generic[_P1, _R1, _P2, _R2]):
+    __wrapped__: Callable[_P2, _R2]
+    def __call__(self, *args: _P1.args, **kwargs: _P1.kwargs) -> _R1: ...
+
+class _Wrapper(Generic[_P1, _R1]):
+    def __call__(self, f: Callable[_P2, _R2]) -> _Wrapped[_P1, _R1, _P2, _R2]: ...
+
+def update_wrapper(
+    wrapper: Callable[_P2, _R2], wrapped: Callable[_P1, _R1], assigned: Sequence[str] = ..., updated: Sequence[str] = ...
+) -> _Wrapped[_P1, _R1, _P2, _R2]: ...
```

Comment:
What motivated this change? `update_wrapper()` modifies `wrapper` and then returns it. Is it to add the `__wrapped__` attribute or to handle cases where `assigned` adds some attributes to `wrapper`?

Reply:
Are you asking specifically about the change for `update_wrapper`? This PR was mostly motivated by a desire to fix `wraps`, but I figured I'd fix `update_wrapper` while I was at it.

The existing definition of `wraps` is pretty broken for a couple of reasons.

First, it leaves the signature of the decorated function unmodified. Typically the decorated function has a signature of `(*args, **kwargs)`, which is quite useless. The purpose of `wraps` is to effectively present a different signature for the decorated function. For example, `@wraps(foo)` should apply the signature of `foo` to the decorated function. This tends to wreak havoc with type checking and completion suggestions. It's one of the most upvoted issues we have with pylance. See https://github.com/microsoft/pylance-release/issues/442 for details.

Second, the existing definition doesn't correctly model the fact that a `__wrapped__` attribute is added to the resulting object. With my proposed change, it not only models the presence of a `__wrapped__` attribute but also provides its correct type (as opposed to a less precise `Callable[..., Any]`).

Reply:
It was specifically asking about `update_wrapper()`. I haven't looked in depth at `wraps()`. (It's a few nestings too deep to think about at 9pm.)

Reply:
For what it's worth, my experience is that making `wraps()` fully transparent itself causes issues: https://github.com/quora/pyanalyze/issues/82. Many wrapping functions modify the signature of the wrapped function.

Reply:
@JelleZijlstra, I'm not clear on what you mean. The link you provided talks about the `@mock.patch` decorator, which is quite different from `@functools.wraps`. How would a wrapping function modify the signature of the wrapped function? If you happen to have an example, that would help me understand.

Reply:
`mock.patch` uses `@wraps` internally. The point here is that many decoorators that use `@wraps` actually don't preserve the signature of the "wrapped" function, so changing the annotations to claim that it does may cause problems.